cpp/include/cugraph/mtmg/per_thread_edgelist.hpp

```diff
@@ -0,0 +1,174 @@
+
+    src_[current_pos_] = src;
+    dst_[current_pos_] = dst;
+    if (wgt) (*wgt_)[current_pos_] = *wgt;
+    if (edge_id) (*edge_id_)[current_pos_] = *edge_id;
+    if (edge_type) (*edge_type_)[current_pos_] = *edge_type;
+
+    ++current_pos_;
+  }
+
+  /**
+   * @brief Append a list of edges to the edge list
+   *
+   * @param handle     The resource handle
+   * @param src        Source vertex id
+   * @param dst        Destination vertex id
+   * @param wgt        Edge weight
+   * @param edge_id    Edge id
+   * @param edge_type  Edge type
+   */
+  void append(handle_t const& handle,
+              raft::host_span<vertex_t const> src,
+              raft::host_span<vertex_t const> dst,
+              std::optional<raft::host_span<weight_t const>> wgt,
+              std::optional<raft::host_span<edge_t const>> edge_id,
+              std::optional<raft::host_span<edge_type_t const>> edge_type)
+  {
+    size_t count = src.size();
+    size_t pos   = 0;
+
+    while (count > 0) {
+      size_t copy_count = std::min(count, (src_.size() - current_pos_));
+
+      std::copy(src.begin() + pos, src.begin() + pos + copy_count, src_.begin() + current_pos_);
+      std::copy(dst.begin() + pos, dst.begin() + pos + copy_count, dst_.begin() + current_pos_);
+      if (wgt)
+        std::copy(wgt.begin() + pos, wgt.begin() + pos + copy_count, wgt_->begin() + current_pos_);
+      if (edge_id)
+        std::copy(edge_id.begin() + pos,
+                  edge_id.begin() + pos + copy_count,
+                  edge_id_->begin() + current_pos_);
+      if (edge_type)
+        std::copy(edge_type.begin() + pos,
+                  edge_type.begin() + pos + copy_count,
+                  edge_type_->begin() + current_pos_);
+
+      if (current_pos_ == src_.size()) { flush(handle); }
+
+      count -= copy_count;
+      pos += copy_count;
+    }
+  }
+
+  /**
+   * @brief Flush thread data from host to GPU memory
+   *
+   * @param handle     The resource handle
+   */
+  void flush(handle_t const& handle)
+  {
+    edgelist_.append(
+      handle,
+      raft::host_span<vertex_t const>{src_.data(), current_pos_},
+      raft::host_span<vertex_t const>{dst_.data(), current_pos_},
+      wgt_ ? std::make_optional(raft::host_span<weight_t const>{wgt_->data(), current_pos_})
+           : std::nullopt,
+      edge_id_ ? std::make_optional(raft::host_span<edge_t const>{edge_id_->data(), current_pos_})
+               : std::nullopt,
+      edge_type_
+        ? std::make_optional(raft::host_span<edge_type_t const>{edge_type_->data(), current_pos_})
+        : std::nullopt);
+
+    current_pos_ = 0;
+  }
+
+ private:
+  detail::per_device_edgelist_t<vertex_t, weight_t, edge_t, edge_type_t>& edgelist_;
+  size_t current_pos_{0};
+  std::vector<vertex_t> src_;
+  std::vector<vertex_t> dst_;
+  std::optional<std::vector<weight_t>> wgt_;
+  std::optional<std::vector<edge_t>> edge_id_;
+  std::optional<std::vector<edge_type_t>> edge_type_;
```

Comment:
Adding {} at the end might be a good habit.

Reply:
Done